src/types/mercuryEvents.ts

```diff
@@ -0,0 +1,156 @@
+import { MercuryAuth } from './auth'
+import { MercurySubscriptionScope } from './subscriptions'
+
+/** An event contract keyed by the event name which also defines the expected body and payload */
+export interface IMercuryEventContract {
+	[eventName: string]: {
+		emitPayload?: Record<string, any>
+		responsePayload?: Record<string, any>
+	}
+}
+
+export type EventSpace = IMercuryEventContract[string]
+
+/** The response from a mercury GQL call */
+export interface IMercuryGQLBody<TBody = Record<string, any>> {
+	data: TBody
+	extensions: {
+		queryCost: number
+		requestMS: number
+		warnings: Record<string, any>[]
+	}
+}
+
+/** Options for mercury.emit() */
+export interface IMercuryEmitOptions<
+	EventContract extends IMercuryEventContract,
+	EventName extends keyof EventContract,
+	EventSpace extends EventContract[EventName]
+> {
+	/** The event name to emit */
+	eventName: EventName
+	/** Optionally specify a unique eventId. This should be a UUID v4 */
+	eventId?: string
+	/** The organization id associated with the event */
+	organizationId?: string | null
+	/** The location id associated with the event */
+	locationId?: string | null
+	/** The user id associated with the event */
+	userId?: string | null
+	/** The event payload */
+	payload?: EventSpace['emitPayload']
+	/** Authenticaiton credentials */
+	credentials?: MercuryAuth
+}
+
+/** Options for mercury.on() */
+export interface IMercuryOnOptions<
+	EventContract extends IMercuryEventContract,
+	EventName extends keyof EventContract,
+	EventSpace extends EventContract[EventName]
+> {
+	/** Whether this handler will provide a respone to the event. */
+	respond?: boolean
+	/** The event to subscribe to */
+	eventName: EventName
+	/** The scope of the data to get back */
+	scope: MercurySubscriptionScope
+	/** A custom UUID for this event. If not provided, one will be generated */
+	eventId?: string
+	/** The organization id where the event is triggered */
+	organizationId?: string | null
+	/** The location id where the event is triggered. If passed, organizationId should also be set. */
+	locationId?: string | null
+	/** The user id who is triggering this event */
+	userId?: string | null
+	payload?: EventSpace['emitPayload']
+	responses?: EventSpace['responsePayload'][]
+}
+
+export interface ISkillOnData {
+	id: string
+	name: string
+	slug: string
+}
+
+export interface IOnData<
+	EventContract extends IMercuryEventContract,
+	EventName extends keyof EventContract,
+	EventSpace extends EventContract[EventName]
+> {
+	/** The event name that is being triggered */
+	eventName: EventName
```

Comment:
I read `IOnData` as anytime data is received, not necessarily that the event has finished. 

but I see it's used to type the actual response of a skill, where maybe naming it something like `IEventResponse` might be a better name since it's technically more accurate (or maybe `IEventSkillResponse`) where `IEventResponse` would be `= IEventSkillResponse[]` Also, putting `Response` in the name fits better with our emit/response payload nomenclature.